XX.md

```diff
@@ -0,0 +1,67 @@
+# NIP XXXX: Read Status Tracking via Salted Hash Keys and Compressed 64-bit Roaring Bitmaps
+
+**Abstract:**  
+Track message read status by publishing replaceable events (kind 30050) containing a compressed 64-bit Roaring Bitmap of read-event keys.
+
+**Specification:**
+
+1. **Secret Salt:**
+  
+  - Generate a random 32-byte salt and encrypt it using NIP‑44.
+  - Include the encrypted salt in every `read_status` event.
+2. **Key Generation:**
+  
+  - For each event, compute its 64-bit key by taking the first 8 bytes of the SHA-256 hash of the concatenated salt and event ID:
+  - ```
+    key = takeFirst8Bytes( sha256( salt + event_id ) )
+    ```
+    
+3. **Roaring Bitmap:**
+  
+  - A compressed data structure for storing integer sets, optimized for fast operations and low memory usage.
+  - In sparse mode, it stores about 2 bytes per entry, versus 32 bytes for raw event IDs, offering >90% size reduction.
+4. **Read Status Events (Kind 30050):**
+  
+  - **Event structure:**
+    
+    ```json
+    {
+      "id": "<event id>",
+      "pubkey": "<publisher pubkey>",
+      "created_at": 1626548400,
+      "kind": 30050,
+      "tags": [
+        ["d", "<sequence identifier>"]
+      ],
+      "content": "{\"bitmap\":\"<base64_encoded_bitmap>\",\"salt\":\"<nip44_encrypted_salt>\"}",
+      "sig": "<signature>"
+    }
+    ```
+    
+  - **Handling large bitmaps:**
+    
+    - Clients can retain the original event and publish new events with only the diff (new read statuses), minimizing network traffic.
+5. **### **Client Behavior:**
+  
+  - **Marking Reads:**  
+    Compute the key for each event and update both the **total read messages bitmap** and the **current read messages bitmap** accordingly.
```

Comment:
This operation isn't clear to me. What exactly is the `update` operation here?
Could you please provide an example on how the encoded bitmap is computed?

Reply:
Normal set operations are possible: add, remove, contains, union, intersection, difference, size, etc.
For code examples I suggest having a look at https://github.com/RoaringBitmap/CRoaring

Typically you would add (mark as read) or remove (mark as unread) a message to your local bitmap, and when you receive a bitmap from an event, you can union the two bitmaps.